### DISCUSSION TITLE
Saving the last directory a Preset file has been exported to via stateData

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#432] [Q&A] [answered]
Hi everybody,
I am improving some of my panels after using for a little while. 
Everytime I save a .syx file from a machine I have to point to the same directory over and over.

I use the following command to open the save file dialog:
`fileToWrite = utils.saveFileWindow("Save preset to disk", File(FileName), "*.syx", true)`

The juce class is "File : https://docs.juce.com/master/classFile.html
[getCurrentWorkingDirectory](https://docs.juce.com/master/classFile.html#a5801c94dd6dad89aeaf6e9ae10b81d9e)
[setAsCurrentWorkingDirectory](https://docs.juce.com/master/classFile.html#af349d3b788121eb6968744e94ff38032)
Those two could do the job but I'm not sure how to pass the last directory I exported a file to, as a string to to a memoryblock.

Is there a way to store the path to a custom userdirectory with the stateData valueTree and saveState/loadState methods?

Thanks a lot for your help ;)

Damien

**Comment by dobo365** [comment]
Salut Damien! J'espère que tu vas bien!

Doing this since long in all my panels ;-)

Indeed, the File object is a gold mine...
For the stateData value tree object you can save anything you want as string. So, just convert to string (if needed) then store

Somewhere in a method in your panel:
```
	-- Save statedata
	sLastFileFullPath = fileToWrite:getFullPathName()
	sLastFileName = fileToWrite:getFileName()
```

SaveStateData method:
```
	stateData:setProperty ("LastFileFullPath", sLastFileFullPath, nil)
	stateData:setProperty ("LastFileName", sLastFileName, nil)
```

ReadStateData method:
```
	sLastFileFullPath = stateData:getProperty("LastFileFullPath")
	sLastFileName = stateData:getProperty("LastFileName")
```

Piece of cake ;-)

**Reply by damiensellier** [reply]
Salut Dominique, oui ca va très bien merci!
unfortunately I already know how to store data in the stateData tree but my main issue is to store the parent folder path and call it back if available.

Juce has getParentDirectory() and it's very handy for that. But my syntax is not good.

**save to file dialog**
```
...
if sUserLastPath ~= nil then
fileToWrite = utils.saveFileWindow("Save preset to disk", File(FileName), sUserLastPath, "*.syx", true) -- syntax to pass the directory is not good
sUserLastPath = fileToWrite:getParentDirectory()
else 
fileToWrite = utils.saveFileWindow("Save preset to disk", File(FileName), "*.syx", true)
sUserLastPath = fileToWrite:getParentDirectory()
end
...
```

For sUserLastPath, the console returns: 
```
console("sUserLastPath : "..tostring(sUserLastPath))
> sUserLastPath : File object: 0x600003226010
```


From the error dialog:
saveFileWindow() requires `(string, File, string, bool)`

The utils.saveFileWindow() function from Ctrlr source https://github.com/RomanKubiak/ctrlr/blob/c1f67d2f4033cb9f5dd20a123854fcddcd99b81d/Source/Lua/CtrlrLuaUtils.cpp#L176 :

According to the Juce documentation, in saveFileWindow("dialog box title", File(), "pattern", bool)
File can either pass the directory or the name of the file, not both.
So I could get the latest directory on mac working but the name has to be written by the user and cannot be automated.

https://github.com/RomanKubiak/ctrlr/blob/c1f67d2f4033cb9f5dd20a123854fcddcd99b81d/JUCE/modules/juce_gui_basics/filebrowser/juce_FileChooser.h#L67

I got the path to directory working this way: 

```
	if sUserLastPath ~= nil then
		fileToWrite = utils.saveFileWindow("Save preset to disk", sUserLastPath, "*.syx", true) -- syntax to pass the directory is good but the filename is missing now
		sUserLastPath = fileToWrite:getParentDirectory()
		console("sUserLastPath after saving : "..(tostring(sUserLastPath)))
	else 
		fileToWrite = utils.saveFileWindow("Save preset to disk", File(FileName), "*.syx", true)
		sUserLastPath = fileToWrite:getParentDirectory()
		console("sUserLastPath after saving : "..(tostring(sUserLastPath)))
	end
```




**saveState**
stateData:setProperty ("UserLastPath", sUserLastPath, nil)


**readState**
sUserLastPath = stateData:getProperty("UserLastPath")


My issue is how to properly pass the last path + the name to the dialog box to avoid navigating to the user folder over and over, not for storing it in the stateData valuetree.

Thanks for your help Dominique, merci d'avance!
Have a nice week

Damien

**Comment by damiensellier** [comment]
Alright so I got it almost working on Mac.
My only doubt is that I'm not sure if it's going to work on PC.

saveFileWindow() requires (string, File, string, bool)
with File the full path to the folder to save into
if File is a fileName the path would be the computer root

So we need the full path + newFileName + .ext

On mac the path are separated with "/" but if I recall on PC the separator for folders is antislash

We need to concatenate the full path like this : 

sUserLastFullPath = tostring(sUserLastPath).."/"..FileName

And this is the result:


```
if sUserLastPath ~= nil then
        console("sUserLastPath : "..tostring(sUserLastPath))
        console("FileName : "..tostring(FileName))
        sUserLastFullPath = tostring(sUserLastPath).."/"..FileName
        console("sUserLastFullPath concat : ".. tostring(sUserLastFullPath))

		fileToWrite = utils.saveFileWindow("Save preset to disk", File(sUserLastFullPath), "*.syx", true)
		sUserLastPath = fileToWrite:getParentDirectory()
		console("sUserLastPath after saving : "..(tostring(sUserLastPath)))
	else 
		fileToWrite = utils.saveFileWindow("Save preset to disk", File(FileName), "*.syx", true)
		sUserLastPath = fileToWrite:getParentDirectory()
		console("sUserLastPath after saving : "..(tostring(sUserLastPath)))
end
```



**the first time the console posts** : 
sUserLastPath : File object: 0x600003ccc870

**the second pass, since sUserLastPath is stored** : 
sUserLastPath : File object: 0x600003cac350
FileName : SPX900 - 27 - ADR-NOISE GATE
sUserLastFullPath concat : File object: 0x600003cac350/SPX900 - 27 - ADR-NOISE GATE
sUserLastPath after saving : File object: 0x600003cb4b60

And an extra .syx extension is added obviously thanks to the 3argument (&pattern)

the stateData has to be stored as string with : 
```
if  sUserLastPath ~= nil then
        sUserLastPath = tostring(sUserLastPath)
        stateData:setProperty("UserLastPath", sUserLastPath, nil)
   end
```

and is shown in the ctrlr.settings like this : 
`<panelCustomData  UserLastPath="File object: 0x600003cb4b60"></panel>`

Can you please let me know if this is working on PC or it requires a antislash ?
Is there a way to tell 'if is mac then...." elseif is PC then" ?

I'm pretty sure there's another way to compose the full path+name in a easier way and avoid this PC/MAC separator problem.
And this path via object ID is not solid, an absolute path like root/folder/folder2/folder3/ would be the way.


Thanks

**Comment by damiensellier** [comment, marked as answer]
Alright I have it and it should work for both PC and Mac.
SInce getParentDirectory() returns an objectID, I prefered using the getFullPathName function, then I remove the last x caracters from the returned string according to the length of the name of the file. This way I have the absolute path until the last folder separator slash for mac or antislash for PC : 

```
	if sLastFileFullPathWoName ~= nil then

        sUserLastFullPathName = sLastFileFullPathWoName..FileName

		fileToWrite = utils.saveFileWindow("Save preset to disk", File(sUserLastFullPathName), "*.syx", true)
		
		sLastFileFullPath = fileToWrite:getFullPathName() -- get the absolute path+filename from dialog box
        sLastFileName = fileToWrite:getFileName() -- get the name from dialog
        sLastFileNameLen = string.len(sLastFileName) -- count the number of caracters in the name from dialog
        sLastFileFullPathWoName = string.sub(sLastFileFullPath, 0, -tonumber(sLastFileNameLen)-1)
               
	else 
		fileToWrite = utils.saveFileWindow("Save preset to disk", File(FileName), "*.syx", true)
		
		sLastFileFullPath = fileToWrite:getFullPathName() -- get the absolute path+filename from dialog box
        sLastFileName = fileToWrite:getFileName() -- get the name from dialog
        sLastFileNameLen = string.len(sLastFileName) -- count the number of caracters in the name from dialog
        sLastFileFullPathWoName = string.sub(sLastFileFullPath, 0, -tonumber(sLastFileNameLen)-1)
	end
	
	console("FileNameFromDialog : "..tostring(sLastFileName))
	console("sLastFileName Length : "..(sLastFileNameLen))
	console("sLastFileFullPath : "..tostring(sLastFileFullPath))
	console("sLastFileFullPathWoName : "..tostring(sLastFileFullPathWoName))
```

the console prints : 

LUA>> FileNameFromDialog : SPX900 - 14 - PERCUSSION ER.syx
LUA>> sLastFileName Length : 31
LUA>> sLastFileFullPath : /Users/Damien/Desktop/SPX900 - 14 - PERCUSSION ER.syx
LUA>> sLastFileFullPathWoName : /Users/Damien/Desktop/

This way we can pass both an automated name for the file and the user last directory. with File(sLastFileFullPathWoName..FileName)

If someone could check if it's working for PC that would be great. 
Et voilà!

**Reply by Godlike-Productions** [reply]
One thing I found on Windows is that the last "/" doesn't get added to the full path name, so you need to add it manually.  I did a check on the right most character.  If it isn't a "/" or "\" I add "/" by concatenating the string.  This should also work with OSX.

I'm busy on some personal court stuff at the moment.  Hit me up with a Message if you need some more help on this and I can post my code from the Jupiter 8 Panel for saving this to the value tree.

**Reply by Godlike-Productions** [reply]
I just re-read your code, and you account for the last character, so ignore my previous comment.  Really elegant solution.  I'll try to find some time to test the code on the PC.

**Reply by dobo365** [reply]
Sorry guys, I still don't understand what the problem was...
Of course that you have to check different things when working with path and filenames on different OS. Seems obvious.
You need to watch on /, \, quotes, spaces...
So important to check/verify what each line is producing

**Reply by damiensellier** [reply]
Hi Dominique, the problem is that the last directory path is returned as an object ID not a string. So my first solution was to concatenate like this (the objectID).."/"..(filename.ext)  but on PC it requires the antislash, so it won't work for PC.
In order to store it and make it work on both platform (win and macOS), the way I found was to get the whole path the last file saved (which is a string) and remove the last file name from it to preserve either the slash or antislash depending on the platform.

**Comment by dnaldoog** [comment]
Hi @damiensellier 

This really is an excellent solution and it works on Windows!

The trouble, as you already found out, is that `static File File::getCurrentWorkingDirectory()` [returns userData](https://docs.juce.com/master/classFile.html#af349d3b788121eb6968744e94ff38032), but not a string unlike say `const String & 	getFullPathName ()`.

 So, as far as I know that userData cannot be output to string and therefore cannot be saved to `function(--[[ ValueTree --]]stateData)`, but you solved that!

I wrote a panel using Atom's code from the DEMO panel and also referring to [this post](https://ctrlr.org/forums/search/getParentDirectory/), which is similar to your code and of course using your code.

I would like to change some of my panels to incorporate this in future.

Excellent work!

----

[remember last directory_1_0_0.zip](https://github.com/RomanKubiak/ctrlr/files/8300168/remember.last.directory_1_0_0.zip)

----

<pre>
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
-- https://ctrlr.org/forums/search/getParentDirectory/

saveContentAsText = function(mod, value)
    if panel:getBootstrapState() then
        return
    end
    if sLastFileFullPathWoName ~= nil then
        local sUserLastFullPathName = sLastFileFullPathWoName
        fileToWrite = utils.saveFileWindow("Save content as " .. ext, File(sUserLastFullPathName), "*" .. ext, true)
    else
        fileToWrite =
            utils.saveFileWindow(
            "Save content as " .. ext,
            File.getSpecialLocation(File.userHomeDirectory),
            "*." .. ext,
            true
        )
    end
    if fileToWrite:isValid() == false then
        return
    end -- Let's see if the file exists
    if fileToWrite:existsAsFile() == false then -- The file is not there, that's ok, let's try to create it
        if fileToWrite:create() == false then -- Ooooh we can't create it, we need to fail here
            utils.warnWindow("File write", "The destination file does not exist, and I can't create it")
            return
        end
    end -- Update the file path
    panel:getLabelComponent("lastFileWritePathL"):setText(" \187 " .. fileToWrite:getFullPathName()) 

    textToWrite = panel:getLabelComponent("saveMe"):getText()
    if textToWrite:length() == 0 then
        utils.warnWindow("Data to write", 'There is no data to write, we\'ll default to "' .. defText .. '"')
        textToWrite = defText
    end 

	-- If we reached this point, we have a valid file we can try to write to
        -- There are two flags when writing text -- asUnicode - if should write our text as UNICODE 
        -- writeUnicodeHeaderBytes - if we should add a special unicode header at the beginning of the file 
        -- we set both to false

    if fileToWrite:replaceWithText(textToWrite, false, false) == false then
        utils.warnWindow("File write", "Failed to write data to file: " .. fileToWrite:getFullPathName())
    end
    local sLastFileFullPath = fileToWrite:getFullPathName() -- get the absolute path+filename from dialog box
    local sLastFileName = fileToWrite:getFileName() -- get the name from dialog
    local sLastFileNameLen = string.len(sLastFileName) -- count the number of characters in the name from dialog
    sLastFileFullPathWoName = string.sub(sLastFileFullPath, 0, -tonumber(sLastFileNameLen) - 1)

end


saveState = function(--[[ ValueTree --]]stateData)

if  sLastFileFullPathWoName ~= nil then
        stateData:setProperty("UserLastPath", sLastFileFullPathWoName, nil)
   end

end

loadState = function(--[[ ValueTree --]] stateData)
if  stateData:hasProperty("UserLastPath") and string.len(stateData:getProperty("UserLastPath")) > 0 then
        sLastFileFullPathWoName=stateData:getProperty("UserLastPath")
   end
end

constructor = function()
defText="no text submitted"
ext="txt"
end
</pre>

**Reply by damiensellier** [reply]
Thanks John, hope it helped, I found this workaround and it worked as a charm. Good to hear it's also working on PC, I was a bit worried about it. Have a nice day!

**Reply by dnaldoog** [reply]
Hi @damiensellier 

I just read that you can get the os path separator (ie is it backslash or forward slash?) with `package.config:sub(1,1)`.

This function also returns the path without the filename. But does lua just use '/' every time anyway internally? Not sure.

<pre>
function getPath(str)
--[[https://stackoverflow.com/questions/9102126/lua-return-directory-path-from-path--]]
local ossep=package.config:sub(1,1)
    --return str:match("(.*[/\\])")
    return str:match("(.*["..ossep.."])")
end
</pre>

**Reply by damiensellier** [reply]
That one seems useful as well. I didn't know you could do that kind of OS specific operations. Thanks John
